### ISSUE TITLE
F821 false positive when deleting inner function after call.

### ISSUE THREAD

**Issue by Skylion007** [astral-sh/ruff#9349]
I recently enabled flake rule F821 on [PyTorch](https://github.com/pytorch/pytorch/pull/116579), but ran into a false positive: https://github.com/pytorch/pytorch/blob/bd10fea79a780b9ba7430c3434849fc7991f6db3/torch/nn/modules/module.py#L2127 . 

It was a bit of a difficult edge case, but I recreated a minimal repro here:
```python
class Dog:
    def f(self, a):
        c = a
        def load(b):
           b+=1

           if b < c+10:
               return load(b)
           return b


        d = load(a)
        del load
        return d
z = Dog().f(2000)
print(z)
assert z == 2010
```
This code is rather unorthodox, but valid. The load function is in scope before it's deleted and therefore the call to the inner recursive load function is defined during the first load call.
